### ISSUE TITLE
abci: Next-generation remote application network protocol

### ISSUE THREAD

**Issue by thanethomson** [cometbft/cometbft#1117]
Relates to #88, and to enable bidirectional requests as per #1112 and #494 in future, the network protocol between remote applications and Comet needs to be reconsidered. The requirements for the future ABCI remote protocol are as follows (will be updated based on feedback on this issue).

The network interaction between a remote application instance and Comet must:
1. Be full-duplex.
2. Offer comparable reliability to TCP.
3. Offer message prioritization whereby messages of higher priority are not (or minimally) blocked by messages of lower priority.
4. Facilitate bidirectional request initiation (i.e. the application must be able to initiate certain requests where Comet responds, and Comet must be able to initiate other requests where the application responds).

Nice-to-haves:
1. Generation of clients and servers should be very easy to support ABCI applications built in any programming language.

The protocol candidates currently under consideration are summarized in the following table.

| Candidate | Full-duplex | Reliable transport | Prioritization | Bidirectional requests | Easy client/server generation |
| --- |:---:|:---:|:---:|:---:|:---:|
| Tendermint Socket Protocol (TSP) | :heavy_check_mark: | :heavy_check_mark: | :white_check_mark: <sup>1</sup>  | :x: | :x: |
| gRPC | :heavy_check_mark: | :heavy_check_mark: | :question: <sup>2</sup>  | :white_check_mark: <sup>3</sup> | :heavy_check_mark: |
| QUIC | :heavy_check_mark: | :heavy_check_mark: <sup>4</sup> | :heavy_check_mark: | :heavy_check_mark: | :x: |
| :question: |  |  |  |  |

_<sup>1</sup> TSP effectively supports a weak form of prioritization through forcing the use of different TCP connections for different types of traffic. It still does not facilitate per-connection prioritization as prioritized framing/chunking of messages would._

_<sup>2</sup> The degree to which we can influence the lower-level HTTP/2 framing and stream prioritization in gRPC communications is unclear._

_<sup>3</sup> Bidirectional gRPC interactions are only either possible through gRPC streaming, which is not really recommended due to potential stability issues, or by way of implementing clients and servers on both the application and Comet, which results in greater configuration and/or initial handshake complexity._

_<sup>4</sup> Assumes interaction on a network where UDP traffic can flow largely unhindered._

**Comment by thanethomson** [comment]
Another option to consider here is [WebTransport](https://developer.chrome.com/articles/webtransport/).
